[DSM] Juniper's kitStatus endpoint PEPPER-841

pegahtah opened the pull request:
## Context

This PR takes care of Juniper's status endpoint.

4 different endpoints and ways of querying DSM
three GET endpoints
* `kitstatus/study/:studyName`
* `kitsstatus/kitId/:kitId`
* `kitstatus/participantId/:participantId`

The forms of POST endpoints
* `kitstatus/kitIds`
      Payload: String[] of kit ids


More can be seen [in this doc](https://docs.google.com/document/d/16o9PiVGYhUusk9rptB55KkjOoj3PioejFYkIzfhSCtc/edit?ouid=117732206884869206335&usp=docs_home&ths=true) 

As per other tickets, both shipKit and kitStatus endpoints are supposed to return either a status in times of success or and error in times of failure of processing a request, so the new implemented class hierarchy for both Juniper endpoints are as below:
![Site map](https://github.com/broadinstitute/ddp-study-server/assets/47120870/448680da-ca4f-49ea-a392-8b7626b5addc)



## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [x] I have considered security and privacy, and added `I-*` labels as needed
- [ ] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [x] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [x] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [x] I have written automated positive tests
- [x] I have written automated negative tests
- [x] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

pegahtah committed:
PEPPER-841 Created StatusKitRoute and NonPepperStatusKitService to return the status of

pegahtah committed:
PEPPER-841 more endpoints to call for kit status

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into PEPPER-841_statusKit

pegahtah committed:
PEPPER-841 added the get and post endpoints, and according methods in KitDao and KitDaoImpl, moved the error strings to KitResponse, created new methods in NonPepperStatusKitService

pegahtah committed:
PEPPER-841 added the new POST endpoint

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into PEPPER-841_statusKit

pegahtah committed:
PEPPER-841
changed JsonTransformer to include null values in response
removed extra space at the end of query conditions in KitDaoImpl
Added more tests in NonPepperKitServiceTest
Corrected some bugs in NonPepperStatusKitService
Added more exception handling

pegahtah committed:
PEPPER-841
Fixed the bug in the POST endpoint where result set was getting closed in StatusKitRoute and NonPepperStatusKitService
 and getKitsByKitId in KitDaoImpl
 Added more tests

pegahtah committed:
PEPPER-841 removed unused coded and renamed code

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDao.java:
```diff
@@ -32,10 +37,18 @@ public interface KitDao extends Dao<KitRequestShipping> {
 
     Optional<ScanError> updateKitLabel(KitRequestShipping kitRequestShipping);
 
+    ResultSet getKitsInDatabaseByInstanceId(DDPInstance ddpInstance);
```
Nit: "InDatabase" can be removed. It is assumed that this is a database query, and this form of naming is inconsistent with the other method names.

pegahtah replied:
So I meant this to show we are getting them from databse and not from ES. Specifically because there is no index for Juniper in ES. I see the redundancy, I will change it

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDaoImpl.java:
```diff
@@ -56,7 +58,7 @@ public class KitDaoImpl implements KitDao {
     public static final String KIT_BY_HRUID = " and bsp_collaborator_participant_id like ? AND not kit_complete <=> 1 "
             + "AND deactivated_date is null";
-
+    private static final Logger logger = LoggerFactory.getLogger(KitDaoImpl.class);
```
Nit: did you intend to move this line?

pegahtah replied:
Yes my checkstyle did not like that... but I checked again and it is not Pepper checkstyle, just Intellij. I will fix it.

denniscunningham replied:
Still not sure why this can't be the first line in the class, which is customary (and as it was previously).

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/KitResponse.java:
```diff
@@ -1,10 +1,11 @@
-    public String errorMessage;
-    public String juniperKitId;
-    public Object value;
+    public static final String ADDRESS_VALIDATION_ERROR = "UNABLE_TO_VERIFY_ADDRESS";
```
Perhaps an enum would be clearer here?

pegahtah replied:
We don't have that many enums in DSM, and we usually keep error strings in static variables. If you think it's necessary I am more than happy to change it, but I'm afraid changing to that can break the consistency. Let me know what you think.

denniscunningham replied:
I'm not sure why enums were not used in other places, but this is a discrete set of constants meant for one purpose, which is what enums are for.
If it gets to be a pain to implement, no big deal, but in a typical Java shop you would get that feedback.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/statics/DBConstants.java:
```diff
@@ -565,4 +565,14 @@ public class DBConstants {
 
     //INDEXES
     public static final String OSTEO_INDEX = "participants_structured.cmi.cmi-osteo";
+    public static final String LABEL_BY = "label_by";
```
Perhaps separate these into a section with its own comment? As is, these look like ES indexes.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/statics/RoutePath.java:
```diff
@@ -31,6 +32,13 @@ public class RoutePath {
     public static final String SHIP_KIT_ENDPOINT = "/shipKit";
+    public static final String STATUS_KIT_ENDPOINT_STUDY = "/statusKit/study/:study";
+    public static final String STATUS_KIT_STUDY = "/statusKit/study/";
+    public static final String STATUS_KIT_ENDPOINT_JUNIPER_KIT_ID = "/statusKit/juniperKitId/:juniperKitId";
```
These would follow the RESTful standard if changed to "/statusKit/juniperKit/:juniperKitId", /statusKit/juniperKit",
   "/statusKit/kits", "/statusKit/participant/:participantId",  "/statusKit/participant"

pegahtah replied:
Good catch!! Thanks! Changed!

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/lddp/util/JsonTransformer.java:
```diff
@@ -1,11 +1,12 @@
 public class JsonTransformer implements ResponseTransformer {
 
-    private Gson gson = new Gson();
+    private Gson gson = new GsonBuilder().serializeNulls().create();
```
Why the change? It seem like JsonTansformer is used widely without this change. IOW, why is `serailzeNulls' needed now?

pegahtah replied:
My first lines of thought was, we should return all of the variables even if they are null, because otherwise it will look like the values for them are just not provided in the response. But now I agree with you about the impact of this file, so I thought with Juniper folks and Brian specifically and we came to the decision that Juniper will assume anything that is not present in the response was null

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/kit/KitDaoMock.java:
```diff
@@ -1,10 +1,15 @@
+import org.broadinstitute.dsm.model.nonpepperkit.NonPepperKitStatus;
+import org.broadinstitute.dsm.model.nonpepperkit.NonPepperStatusKitService;
 
 public class KitDaoMock implements KitDao {
```
Are we actually using this? If not, we should get rid of it. Since all methods do nothing is does not appear useful.

pegahtah replied:
Oh I would love to remove this!!

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/JuniperKitStatusTest.java:
```diff
@@ -0,0 +1,171 @@
+    private Config cfg;
+
+    @Before
+    public void beforeClass() {
```
It appears that all the tests in this file are ignored?
I suggest not using the TestHelper.setupDB pattern since it typically works against the dev DB, and it is unclear how this would work during CI. Also, we will probably get rid of this pattern entirely at some point.

Instead tests using the DB should instead extend `DbTxnBaseTest`. (See `DbTxnBaseTest` uses in DSM to see how this works.)

pegahtah replied:
I am working on that in this same PR with commits as part of PEPPER-1052

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/JuniperKitCreationTest.java:
```diff
@@ -53,20 +54,16 @@ public void createNewMockJuniperKitTest() {
         Assert.assertEquals(newKits.size(), oldkits.size() + 1);
-        KitRequestShipping newKit = newKits.stream().filter(kitRequestShipping -> kitRequestShipping.getDdpParticipantId().equals(participantId + rand)).findAny()
-                .get();
+        KitRequestShipping newKit =
```
See my comment for JuniperKitStatusTest. Same applies here.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/juniper/StatusKitRoute.java:
```diff
@@ -0,0 +1,51 @@
+            log.info(String.format("Got a request to return information of kits in non-pepper study %s", study));
+            return this.nonPepperStatusKitService.getKitsBasedOnStudyName(study);
+        }
+        if (request.url().contains(RoutePath.STATUS_KIT_JUNIPER_KIT_ID)) {
```
Should each of these cases be `else if`? 
Also, should there be a final `else` to cover any path not covered here? What happens when `null` is returned? What does the HTTP response look like in that case?
Does it make sense to check the request type (GET, POST, etc.)?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/user/UserDao.java:
```diff
@@ -141,4 +145,31 @@ public int delete(int id) {
         return (int) simpleResult.resultValue;
     }
+
+    public static HashMap<Integer, UserDto> selectAllUsers() {
```
This should return `Map` not `HashMap`.
Since this method is public and returns something non-obvious, it should probably have JavaDoc explaining what it returns.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/user/UserDao.java:
```diff
@@ -141,4 +145,31 @@ public int delete(int id) {
+        });
+
+        if (results.resultException != null) {
+            throw new RuntimeException("Error getting list of all users ", results.resultException);
```
Throw DsmInternalError instead of RunTimeException.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDaoImpl.java:
```diff
@@ -666,4 +681,73 @@ public boolean hasKitReceived(Connection connection, String ddpParticipantId) {
+            return dbVals;
+        });
+        if (simpleResult.resultException != null) {
+            throw new DSMBadRequestException(simpleResult.resultException);
```
This should probably be DsmInternalError?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDaoImpl.java:
```diff
@@ -666,4 +681,73 @@ public boolean hasKitReceived(Connection connection, String ddpParticipantId) {
+            return dbVals;
+        });
+        if (simpleResult.resultException != null) {
+            throw new DSMBadRequestException(simpleResult.resultException);
```
This should probably be DsmInternalError?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDaoImpl.java:
```diff
@@ -666,4 +681,73 @@ public boolean hasKitReceived(Connection connection, String ddpParticipantId) {
+            return dbVals;
+        });
+        if (simpleResult.resultException != null) {
+            throw new DSMBadRequestException(simpleResult.resultException);
```
This should probably be DsmInternalError?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDaoImpl.java:
```diff
@@ -666,4 +681,73 @@ public boolean hasKitReceived(Connection connection, String ddpParticipantId) {
+
+    @Override
+    public ArrayList<NonPepperKitStatus> getKitsByKitIdArray(String[] kitIdsArray, NonPepperStatusKitService nonPepperStatusKitService) {
+        ArrayList<NonPepperKitStatus> list = new ArrayList<>();
```
How large is this array typically? If large, perhaps a single query with post-processing would be more efficient?

pegahtah replied:
We decided it should not be larger than 20, if they need more, they will send the rest in another request

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDaoImpl.java:
```diff
@@ -666,4 +681,73 @@ public boolean hasKitReceived(Connection connection, String ddpParticipantId) {
         return false;
     }
+
+    public ResultSet getKitsInDatabaseByInstanceId(DDPInstance ddpInstance) {
```
From a separation of concerns point of view, it would be better to return DTOs instead of result sets.
Is there a reason you chose this approach?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperStatusKitService.java:
```diff
@@ -0,0 +1,149 @@
+        String discardDate = convertTimeStringIntoTimeStamp(foundKitResults.getLong(DBConstants.DISCARD_DATE));
+
+        NonPepperKitStatus nonPepperKitStatus =
+                new NonPepperKitStatus(foundKitResults.getString(DBConstants.DDP_KIT_REQUEST_ID),
```
This is difficult to read and seems brittle. Some other choices: use a builder pattern, make the NonPepperKitStatus fields public and assign directly.
I realize the benefit of this approach is that every field is accounted for, but it's really difficult to read this code and determine that the current argument list is correct.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperStatusKitService.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    private NonPepperKitStatus createStatusForKit(ResultSet foundKitResults, HashMap<Integer, UserDto> users) throws SQLException {
+        String labelUser = getUserEmailForFields(foundKitResults.getString(DBConstants.LABEL_BY), users);
```
As mentioned in my comments for the DAO, it feels like this code belongs in the DAO.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperStatusKitService.java:
```diff
@@ -0,0 +1,149 @@
+
+    public KitResponse getKitsBasedOnStudyName(String studyGuid) {
+        if (StringUtils.isBlank(studyGuid)) {
+            return new KitResponseError(KitResponse.MISSING_STUDY_GUID);
```
This would typically be a 400 error, but it looks like it probably returns a 200?
Likewise for the following checks.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperKitCreationService.java:
```diff
@@ -24,39 +24,33 @@
     public KitResponse createNonPepperKit(JuniperKitRequest juniperKitRequest, String studyGuid, String kitTypeName) {
         if (StringUtils.isBlank(juniperKitRequest.getJuniperParticipantID())) {
-            return new KitResponse(MISSING_JUNIPER_PARTICIPANT_ID, juniperKitRequest.getJuniperKitId(),
+            return new KitResponseError(KitResponse.MISSING_JUNIPER_PARTICIPANT_ID, juniperKitRequest.getJuniperKitId(),
```
This would typically be a 400 error, but it looks like it probably returns a 200?
Likewise for the following checks.

pegahtah replied:
In `JuniperShipKitRoute` There is a check for the response of this method
```
if (!(kitResponse instanceof KitResponseError)) {
            response.status(200);
        } else {
            response.status(400);
        }
```

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperStatusKitService.java:
```diff
@@ -0,0 +1,149 @@
+            log.error(studyGuid + " is not a Juniper study!");
+            return new KitResponseError(KitResponse.UNKNOWN_STUDY);
+        }
+        BookmarkDto kitPagingThreshold = new BookmarkDao().getBookmarkByInstance(STATUS_KIT_THRESHOLD)
```
This appears to be unused?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/juniper/StatusKitRoute.java:
```diff
@@ -0,0 +1,51 @@
+    @Override
+    public Object handle(Request request, Response response) throws Exception {
+
+        if (request.url().contains(RoutePath.STATUS_KIT_STUDY)) {
```
It does not look like we check user access at any point?

pegahtah replied:
Same as all of the other non UI  endpoints, this is getting passed by authenticating their secret instead of their access roles

pegahtah replied:
Currently `before(apiRoot + "*", new LoggingFilter(auth0Domain, auth0claimNameSpace, bspSecret, null, bspSecretEncoded));` in DSMServer takes care of this endpoint and also all of the GP endpoints. I have a ticket to differentiate between the two services (GP and Juniper) in time of authentication so each can use their own unique secret

denniscunningham replied:
Learned something!
So it basically works that if you have the token you can do all the operations for all the affected endpoints? IOW, no special roles?
Is that likely to be true for the long term of this service? (Not that anything is required now, just curious.)

pegahtah replied:
Yes no special roles are defined for this in database, and also not for Mercury or BSP. We do check the token signature and expiration time and the secret matching the one we have on file.
I think it is, not sure what long term means anymore for DSM, but yes I don't think there is any ticket or talk to change that. If you want we can discuss this more and make plans

denniscunningham reviewed (changes requested):
I may have more comments, but there are a lot here now which may change some of the code. So here's what I have so far.

pegahtah committed:
PEPPER-841 reverted back JsonTransformer class , changed endpoints names to have RESTful standards,created a KitStatusDao class and moved my changes from KitDao into that, also deleted KitDaoMock.java
changed Strings in KitResponse to enums, added checks for request methods in StatusKitRoute, changed NonPepperKitStatus to NonPepperKitStatusDto

pegahtah committed:
PEPPER-1052
Added JuniperSetupUtil with methods to create a fake Juniper study and to delete all of the added data later. made changes to KitStatusTest to ame sure it can still run

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into PEPPER-841_statusKit

pegahtah committed:
PEPPER-1052 fixed JuniperKitCreationTest to inherit DbTxnBaseTest

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into PEPPER-841_statusKit

pegahtah committed:
PEPPER-1052 tweaked main-config.yml to have DSM config in it as well

pegahtah committed:
PEPPER-1052 changed secret name

pegahtah committed:
PEPPER-1052 changed main-config.yml

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into PEPPER-841_statusKit

pegahtah committed:
PEPPER-1052 adding env variable

pegahtah committed:
PEPPER-1052 removed env variable

pegahtah committed:
PEPPER-1052 adding PEPPER-* to branch name possibilities

pegahtah committed:
PEPPER-1052 fixed java code to get the VM variable

pegahtah committed:
PEPPER-1052 added instructions on how to run tests

pegahtah committed:
PEPPER-1023 changing the docker image version to use in main-config.yml

pegahtah committed:
PEPPER-1023 checking if config is being read appropriately

pegahtah committed:
PEPPER-1052 changing the secret ci gets

pegahtah committed:
PEPPER-1052 adding more logs and more config reads

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into PEPPER-841_statusKit

pegahtah committed:
PEPPER-1052 moved checkAddress to from NonPepperKitCreationService to EasyPostUtil, changed pom to add Mockito and used Mockito to mock EasyPostUtil, removed reading dsm secrets from both main-config.yml and JuniperSetupUtil

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into PEPPER-841_statusKit

pegahtah committed:
PEPPER-1052 checking something in ci

pegahtah committed:
PEPPER-1052 moving query into classes

pegahtah commented on .circleci/main-config.yml:
```diff
@@ -128,6 +128,9 @@ commands:
+            ;;
             ?(*_)test)
               DEPLOY_ENV=test
             ;;
```
General comment: I kept what I thought have to be fixed in this file but removed my DSM secret changes

denniscunningham replied:
Can you explain why this change is needed?

pegahtah replied:
I'm following the pattern here. I noticed that branches labeled `DDP-*` can be deployed to the dev environment. However, it seems that the `DDP-` naming convention was used before we transitioned to being the Pepper swarm and updated our Jira ticket naming from `DDP-` to `PEPPER-*`. Considering this change in naming, I'm concerned that we might have lost some functionality during this transition. To address this, I went ahead and included the `PEPPER-*` branches for deployment as well. This is to ensure that we're not missing any relevant changes that were made after the naming switch.

denniscunningham replied:
Thanks for the explanation. Let me see if I understand...
This script supports deployment of our RC branches? And those are typically titled like `dsm_rc_swarm_sprint_12`, although hotfix branches likely have different names. But since that branch name does not appear to match any of these cases, my mental model must be wrong. What am I missing?

Also, how did we choose a new test executor image? It looks like that image was built in March. Do we know why it was not put into use then?
`

pegahtah replied:
I think I now realize your point. The `rc` and `hotfix`  name patterns that trigger storing the jar are actually checked in the config.yml file. I also see more `ddp_*` patterns in that one, which at this point makes me second guess my changes to the main-config.yml in this PR. Since DSM was not using circle for deployment before last year, I will revert this change back for now. Maybe we can check this later. 

I still think we need the newer image though

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:
```diff
@@ -1294,8 +1300,7 @@ public static int getKitCounter(@NonNull Connection conn, String collaboratorSam
     public static String generateDdpLabelID() {
-        return generateDdpLabelID(15,
-                DSMConfig.getSqlFromConfig(ApplicationConfigConstants.GET_FOUND_IF_KIT_WITH_DDP_LABEL_ALREADY_EXISTS));
+        return generateDdpLabelID(15, GET_FOUND_IF_KIT_WITH_DDP_LABEL_ALREADY_EXISTS);
```
Moved these two queries from conf file into class, it's a pattern we have been following for the last years to move queries into classes when we change the classes

denniscunningham replied:
So they do not need to be in config? IOW, they are static?

pegahtah replied:
Yes, we started making them static as an overall rule a couple of years ago. To name one issue, having all of them in the config reduced code readability and made it hard to debug and track the impacts of our changes

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperKitCreationService.java:
```diff
@@ -89,59 +73,13 @@ public KitResponse createNonPepperKit(JuniperKitRequest juniperKitRequest, Strin
-        }
-        log.info("Address is not valid " + juniperKitRequest.getShortId());
-        return false;
-
```
Moved this into EasyPostUtil class

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/JuniperShipKitRoute.java:
```diff
@@ -1,51 +0,0 @@
-    public Object handle(Request request, Response response) throws Exception {
-        return createNonPepperKit(request, response);
-    }
-}
```
Moved class into `juniper` package

pegahtah committed:
PEPPER-1052 removing old comment

denniscunningham commented on pepper-apis/dsm-core/pom.xml:
```diff
@@ -147,6 +147,13 @@
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <version>5.4.0</version>
```
This dependency is already defined in this POM and the parent. 
If you need to update the version, that happens in the parent.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/DSMServer.java:
```diff
@@ -96,7 +97,7 @@
 import org.broadinstitute.dsm.route.FilterRoute;
 import org.broadinstitute.dsm.route.InstitutionRoute;
-import org.broadinstitute.dsm.route.JuniperShipKitRoute;
+import org.broadinstitute.dsm.route.juniper.JuniperShipKitRoute;
```
Are you getting a lint warning on this include? Seems like it is out of order.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/DSMServer.java:
```diff
@@ -588,6 +590,13 @@ protected void setupCustomRouting(@NonNull Config cfg) {
         if (!cfg.getBoolean("ui.production")) {
             NonPepperKitCreationService nonPepperKitCreationService = new NonPepperKitCreationService();
             post(apiRoot + RoutePath.SHIP_KIT_ENDPOINT, new JuniperShipKitRoute(nonPepperKitCreationService), new JsonTransformer());
```
At this point I am unclear about why the service is created here instead of in the route handler. Perhaps it will be clearer when I look at the route code.

pegahtah replied:
I changed this and moved them into the route.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:
```diff
@@ -197,6 +197,13 @@ public class KitRequestShipping extends KitRequest implements HasDdpInstanceId {
+            " SELECT count(*) kitRequestCount from ddp_kit_request where bsp_collaborator_sample_id REGEXP \"^%1\" and kit_type_id = ?";
+
+    private static final String GET_FOUND_IF_KIT_WITH_DDP_LABEL_ALREADY_EXISTS =
+            " select 1 as found  from (select 1 from ddp_kit_request req  where req.ddp_label = ?) as existing_rows ";
```
I'm not following the logic of this statement. Is this insufficient: `select 1 from ddp_kit_request req  where req.ddp_label = ?`?

pegahtah replied:
Looking at it again, no I think that should be fine... 🤔 not sure why it was made like that from the beginning

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:
```diff
@@ -1001,9 +1008,9 @@ public static String writeRequest(@NonNull String instanceId, @NonNull String dd
                     //This error will trigger on studies with no participants, this skips
                     //the error log if that is the reason for the upsert failure.
-                        logger.error(String.format("Error inserting newly created kit request shipping with dsm kit request id: %s in "
+                    logger.error(String.format("Error inserting newly created kit request shipping with dsm kit request id: %s in "
```
It feels like this method should be more generic, and this error handling should be in the caller. (The method comment says it is called by a chron job, but this can easily be decoupled from that. IOW, this will be easier to test/verify if it does not make assumptions about the callers.)

pegahtah replied:
To my understanding, this method takes care of writing the new kit ( from any study) into DB, which needs a number of queries to execute in one transaction, hence the length of the method, and then, if db update and insertion was successful, we also update ES. 

Looking at the history of this file, this comment was made for Darwin's arc because it does not have an es participants index. However, now that I have added a check for juniper ( and juniper-like studies that do not have an es participants index, which is also true for Darwin's) that comment is not valid anymore and this catch is only entered through the point of a real ES update failure. 

I may be missing the point of handling the error in the caller here not that the comment is not valid anymore. Please let me know if I am not looking at it the right way

denniscunningham replied:
Sorry, my comment was not clear.
I'm concerned about converting the exception to a log statement. Generally we would only do that at the interface point -- for example, http request handling, or pusbsub message handling, etc. This method appears to be called from a handful of different places, including a route. In that case the right thing to is let the callers handle the exception as they see fit, rather than swallowing it here.
I realize the form of this code already existed, and this is an opportunity for improvement. But I appreciate that is not technically in scope for this PR.

pegahtah replied:
I see what you mean... I completely agree... My personal thought is that would go beyond the scope of Juniper status route... However, if you think this is needed I can make the changes. I just don't think changing all of those kit-related classes will benefit this specific PR. How about I make a tech-debt ticket and address them in there?

denniscunningham replied:
Sounds good.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:
```diff
@@ -197,6 +197,13 @@ public class KitRequestShipping extends KitRequest implements HasDdpInstanceId {
             "update ddp_kit_request set order_transmitted_at = ? where external_order_number = ?";
+
+    private static final String GET_COUNT_KITS_WITH_SAME_COLLABORATOR_SAMPLE_ID_AND_KIT_TYPE =
+            " SELECT count(*) kitRequestCount from ddp_kit_request where bsp_collaborator_sample_id REGEXP \"^%1\" and kit_type_id = ?";
```
Are SM IDs actually regular expressions?

pegahtah replied:
I think you mean collaborator sample ids, they are made like `<project name>_<short id>_<sample type>_<counter>`.  SM IDs are barcodes assigned to tissue samples by study staff or GP, I _think_ they are made like `SM-<some number>`

pegahtah replied:
So here we check for if there are any samples with collaborator sample id starting with `<project name>_<short id>_<sample type>` to see what the counter should be

denniscunningham replied:
But is the replacement for `%1` actually a regular expression? If not we should get rid of `REGEXP` in the query since it can only cause subtle bugs if a reg expression character ever appears in a collaborator sample ID.

pegahtah replied:
Yes it starts with `^`

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitDaoImpl.java:
```diff
@@ -96,18 +94,20 @@ public class KitDaoImpl implements KitDao {
+                    + "LEFT JOIN ddp_kit_tracking AS track ON track.kit_label = ?"
+                    + "LEFT JOIN kit_type AS kt ON kt.kit_type_id = req.kit_type_id "
+                    +
+                    "LEFT JOIN ddp_kit_request_settings AS ks ON ks.kit_type_id = req.kit_type_id AND ks.ddp_instance_id = req.ddp_instance_id "
```
There is a lot to compare here, but this looks like just a reformatting?

pegahtah replied:
Yes, line length got fixed

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitStatusDao.java:
```diff
@@ -0,0 +1,164 @@
+    private static final String SELECT_KIT_STATUS =
+            " SELECT req.*, k.*, discard.*, tracking.tracking_id as return_tracking_number, tracking.scan_by as tracking_scan_by, "
+                    + " tracking.scan_date as tracking_scan_date "
+                    + " FROM ddp_kit_request req LEFT JOIN ddp_kit k on (k.dsm_kit_request_id = req.dsm_kit_request_id) "
```
Putting the join on a separate line makes this easier to read.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitStatusDao.java:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    public List<NonPepperKitStatusDto> getKitsByInstanceId(DDPInstance ddpInstance, HashMap<Integer, UserDto> users) {
+        ArrayList<NonPepperKitStatusDto> list = new ArrayList<>();
```
This should be `List<NonPepperKitStatusDto> list...`. Same with the similar methods below. The linter should be catching this.

pegahtah replied:
Interesting, checkstyle is not finding any problems with this file... I will make the changes to more general returns

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitStatusDao.java:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    private static class BuildNonPepperKitStatusDto {
+        public Object build(ResultSet foundKitResults, HashMap<Integer, UserDto> users) throws SQLException {
```
Why does this return an Object? Seems like NonPepperKitStatusDto would be a better return type.

denniscunningham replied:
Perhaps this method should catch the SQLException and return a DsmInternalError? Callers probably don't know what to do with a SQLException, but should know how to handle a DsmInternalError.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/user/UserDao.java:
```diff
@@ -141,4 +146,36 @@ public int delete(int id) {
+     * @return Map instance, key is the user id and value is UserDto
+     * */
+    public static Map<Integer, UserDto> selectAllUsers() {
+        HashMap<Integer, UserDto> users = new HashMap<>();
```
The type should be Map, not HashMap.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/KitResponse.java:
```diff
@@ -1,10 +1,23 @@
-    public String juniperKitId;
-    public Object value;
+    public enum UsualErrorMessage {
+        ADDRESS_VALIDATION_ERROR("UNABLE_TO_VERIFY_ADDRESS"),
```
I believe enums have a name() method, so UNKNOWN_STUDY("UNKNOWN_STUDY") can be just UNKNOWN_STUDY. But it looks like you want the flexibility to change the message, since ADDRESS_VALIDATION_ERROR does not fit the pattern?

pegahtah replied:
yes, I was thinking there is a chance we might make this errors look/sound sophisticated after the first dry run and integration test. I don't have any personal preference here

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/KitResponse.java:
```diff
@@ -1,10 +1,23 @@
-    public String errorMessage;
-    public String juniperKitId;
-    public Object value;
+    public enum UsualErrorMessage {
```
Could this enum just be part of KitResponseError or a standalone enum?

denniscunningham replied:
Probably could simply name this ErrorMessage? Unclear what the term 'Usual' adds.

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/JuniperSetupUtil.java:
```diff
@@ -0,0 +1,325 @@
+import org.broadinstitute.dsm.util.DSMConfig;
+import org.broadinstitute.lddp.db.SimpleResult;
+
+@Slf4j
```
A class comment would be helpful.

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/JuniperSetupUtil.java:
```diff
@@ -0,0 +1,325 @@
+                    throw new RuntimeException("Error inserting in ddp_instance_group ", e);
+                }
+            } catch (SQLException ex) {
+                dbVals.resultException = ex;
```
This is not useful. This error will just be lost.
Assuming this code will only be used to setup tests, I suggest not catching these exceptions -- just let them bubble up to the test methods. Or you could create a method that sits on top of this and consolidates the exceptions into a DsmInternalError.

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/JuniperSetupUtil.java:
```diff
@@ -0,0 +1,325 @@
+                            ddpGroupId = rs.getString(1);
+                        }
+                    } catch (Exception e) {
+                        throw new RuntimeException(e);
```
A couple things: 
Whenever possible do not throw RuntimeExceptions. DSM has more descriptive exception classes. (Lint should be providing that warning.)
There is no reason to catch and re-throw exceptions if this is just for testing. But it you want to catch an re-throw it makes more sense to put a method above this to do that. All these re-throws just make the code hard to read.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperKitCreationService.java:
```diff
@@ -19,44 +19,30 @@
+                                          DDPInstance ddpInstance) {
         if (StringUtils.isBlank(juniperKitRequest.getJuniperParticipantID())) {
-            return new KitResponse(MISSING_JUNIPER_PARTICIPANT_ID, juniperKitRequest.getJuniperKitId(),
+            return new KitResponseError(KitResponse.UsualErrorMessage.MISSING_JUNIPER_PARTICIPANT_ID.getMessage(),
```
See my comment in KitReponse, but passing the enum without the getMessage would probably be simpler. If you keep the getMessage part it can be buried in the KitResponseError constructor.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperStatusKitService.java:
```diff
@@ -0,0 +1,101 @@
+        HashMap<Integer, UserDto> users = (HashMap<Integer, UserDto>) UserDao.selectAllUsers();
+        // get the kit with that juniperKitId
+        ArrayList<NonPepperKitStatusDto> list = (ArrayList<NonPepperKitStatusDto>) kitStatusDao.getKitsByJuniperKitId(juniperKitId, users);
+        StatusKitResponse statusKitResponse = new StatusKitResponse(list);
```
No need for this assignment. Lint should catch this.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/juniper/StatusKitRoute.java:
```diff
@@ -0,0 +1,59 @@
+    @Override
+    public Object handle(Request request, Response response) throws Exception {
+        KitResponse kitResponse;
+        if (request.requestMethod().equals(RoutePath.RequestMethod.GET.toString()) && request.url().contains(RoutePath.KIT_STATUS_STUDY)) {
```
Seems like this logic could be simplified. This statement is repeated as part of several AND statements: `request.requestMethod().equals(RoutePath.RequestMethod.GET.toString()`. The simpler version does this test once, then does the other conditionals within that branch.

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/JuniperSetupUtil.java:
```diff
@@ -0,0 +1,325 @@
+    private static String carrierId;
+    private static String ddpKitRequestSettingsId;
+
+    public static void setupUpAJuniperInstance(String instanceName, String studyGuid, String displayName, String collaboratorPrefix) {
```
This method name is non-standard: It repeats 'up', and as a rule method names (and the names of pretty much anything in Java) do not include articles ('A', 'The', etc.). It seems like 'setupJuniperInstance' would work here?

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/JuniperSetupUtil.java:
```diff
@@ -0,0 +1,325 @@
+        }
+    }
+
+    public static void deleteJuniperKit(String ddpKitRequestId) {
```
I would expect an inverse method to 'setupJuniperInstance' in this class, like 'deleteJuniperInstance'. Does this method delete things created elsewhere? If so, where do the Juniper instances created above get deleted?

pegahtah replied:
If I understood you correctly, yes, see `deleteJuniperTestStudies()` above this

denniscunningham replied:
From the names, it sounds like these methods are dealing with different entities, instances vs studies. It does not look like `deleteJuniperTestStudies` actually deletes any studies. A reader (or user) of this class would typically look for `setup<term>` and `delete<term>` methods. Kind of a nit, but without it the class seems more like a work in progress than finished.
Does that make sense?

pegahtah replied:
I don't think I understand totally
Deleting a study would need deleting all of the configs and settings set for the study as well. For the clean removal of foreign keys, it should be in the order it is done in that method. 
Considering the setup is called `setupJuniperInstance`, does `deleteJuniperInstanceAndSettings()` sound better?

denniscunningham replied:
Presumably you created this class to encapsulate the setup and teardown of data for testing. As a caller I want to know what to call to set things up and what to call to tear things down. Ideally that is all self-contained here: everything created by a single method is torn down by a different single method. The tear down should not consider any setup other that what is setup here.
Given that, as a caller I would expect to see at setupXXX method and a deleteXXX method. Anything other than that means I have to closely inspect the code to see what the code is doing other than what I described above.

Does that make sense? 
If so, and the delete tears down exactly what the setup sets up, then I suggest either setupJuniperInstance and deleteJuniperInstance, or setupJuniperInstanceAndSettings and deleteJuniperInstanceAndSettings.

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/JuniperSetupUtil.java:
```diff
@@ -0,0 +1,325 @@
+        });
+    }
+
+    public static void loadDSMConfig() {
```
How does this work if the created object is not returned?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperStatusKitService.java:
```diff
@@ -0,0 +1,101 @@
+        this.kitStatusDao = new KitStatusDao();
+    }
+
+    public static String getUserEmailForFields(String userIdInDB, HashMap<Integer, UserDto> users) {
```
What does this method do?

pegahtah replied:
Juniper wants to know who did the scan, sent, received and deactivation of the kit. DSM keeps those as user ids in the kit tables. Also sometimes those are written as "SYSTEM". This method _tries_ to match the user id to an email , but if the id is not an integer ( like the `SYSTEM` I mentioned) it just returns that

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperStatusKitService.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    public static String convertTimeStringIntoTimeStamp(Long dateMillis) {
+        if (dateMillis == null || dateMillis == 0L) {
```
Should this be an assert? IOW, should this throw a DsmInternalError when it gets bad input, like an assert? What will the caller do with the NULL return? 
It feels like the only upside of this method would be the assert, and it is not present.

pegahtah replied:
This is for the fields that do not have a data yet, for example if there is not a scan date, return null. Did I understand your question and does that make sense?

denniscunningham replied:
Yes, thank you. It looks like this is used when building a DTO, so as long as the underlying columns in the DB can handle null this is fine.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/juniper/StatusKitRoute.java:
```diff
@@ -0,0 +1,59 @@
+            kitResponse = this.nonPepperStatusKitService.getKitsBasedOnParticipantId(participantId);
+        } else if (request.requestMethod().equals(RoutePath.RequestMethod.POST.toString())
+                && request.url().contains(RoutePath.KIT_STATUS_ENDPOINT_KIT_IDS)) {
+            String[] kitIds = new Gson().fromJson(request.queryMap().get(RoutePath.JUNIPER_KIT_IDS).value(), String[].class);
```
Seems like this can also throw a JSON parsing exception?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperStatusKitService.java:
```diff
@@ -0,0 +1,101 @@
+        return instant.toString();
+    }
+
+    public KitResponse getKitsBasedOnStudyName(String studyGuid) {
```
Nit: 'BasedOn' could be replaced by the simpler 'By'.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperStatusKitService.java:
```diff
@@ -0,0 +1,101 @@
+        }
+        DDPInstance ddpInstance = DDPInstance.getDDPInstanceWithRoleByStudyGuid(studyGuid, "juniper_study");
+        if (ddpInstance == null || !ddpInstance.isHasRole()) {
+            log.error(studyGuid + " is not a Juniper study!");
```
Do we want this to be a production alerting error? Seems like bad user input.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitStatusDao.java:
```diff
@@ -0,0 +1,164 @@
+                    .withLabelDate(
+                            NonPepperStatusKitService.convertTimeStringIntoTimeStamp(foundKitResults.getLong(DBConstants.LABEL_DATE)))
+                    .withLabelByEmail(
+                            NonPepperStatusKitService.getUserEmailForFields(foundKitResults.getString(DBConstants.LABEL_BY), users))
```
This join would typically be done in the query.

pegahtah replied:
I'd agree and made it that way, but here's what I thought:

kits have user ids for creation, scan, tracking scan, received and deactivated. That would mean 5 extra conditions or joins in the query (based on how we write it), meanwhile we only have around 150 real users in dsm so we can just query them and select from that map after that. I thought this is actually less time consuming and more light weight for the server.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dto/kit/nonPepperKit/NonPepperKitStatusDto.java:
```diff
@@ -0,0 +1,186 @@
+
+@Getter
+public class NonPepperKitStatusDto {
+    String juniperKitId;
```
These member vars should probably all be private.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/exception/DSMBadRequestException.java:
```diff
@@ -5,4 +5,7 @@ public DSMBadRequestException(String message) {
         super(message);
     }
 
+    public DSMBadRequestException(Exception e) {
```
This appears to be unused.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/juniper/JuniperShipKitRoute.java:
```diff
@@ -0,0 +1,73 @@
+
+    NonPepperKitCreationService kitCreationService;
+
+    public JuniperShipKitRoute(NonPepperKitCreationService kitCreationService) {
```
Why is this instance passed in instead of created in this class? It does not appear expensive to instatiate.

pegahtah replied:
I was worried of the load, will change it

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/juniper/JuniperShipKitRoute.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    public KitResponse createNonPepperKit(Request request, Response response) {
+        ShipKitRequest shipKitRequest = new Gson().fromJson(request.body(), ShipKitRequest.class);
```
The exception thrown here should be caught and handled.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/juniper/JuniperShipKitRoute.java:
```diff
@@ -0,0 +1,73 @@
+        String studyGuid = shipKitRequest.getJuniperStudyGUID();
+        DDPInstance ddpInstance = DDPInstance.getDDPInstanceWithRoleByStudyGuid(studyGuid, "juniper_study");
+        if (ddpInstance == null) {
+            log.error(studyGuid + " is not a study!");
```
Here, and below, it appears we are creating production alerts for bad user input?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/juniper/JuniperShipKitRoute.java:
```diff
@@ -0,0 +1,73 @@
+            return new KitResponseError(KitResponse.UsualErrorMessage.UNKNOWN_STUDY.getMessage(), juniperKitRequest.getJuniperKitId(),
+                    studyGuid);
+        }
+        if (!ddpInstance.isHasRole()) {
```
I'm puzzled by hasRole... The code is not clear on what this is, and the DDPInstance code is pretty confusing. I realize this problem was not introduced by this code, but in those cases -- where we are dealing with poorly implemented underlying code -- a comment explain what is going on is helpful.

pegahtah replied:
Will make comment in the code, but here a little explanation as a background:

As you know, Pepper is very complex in terms of different studies having different capabilities. For all of those capabilities, we have an `instance_role` table, in which we have all of the different functionalities a study can have in DSM. In `ddp_instance_role` we can assign or take away "roles" (which are really just functionalities) from studies. When we call `getDDPInstanceWithRoleByStudyGuid(studyGuid, role)` we get a boolean value that shows if the study also is set to use this role. I want to make sure this endpoint is not creating a new kit for a study that is not designed to use this method.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/juniper/JuniperShipKitRoute.java:
```diff
@@ -0,0 +1,73 @@
+@Slf4j
+public class JuniperShipKitRoute implements Route {
+
+    NonPepperKitCreationService kitCreationService;
```
This can be private and final

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/juniper/JuniperShipKitRoute.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Override
+    public Object handle(Request request, Response response) throws Exception {
```
Ideally this does *not* throw an exception -- the route code has caught and handled those -- but the signature is inconsistent with createNonPepperKit. I suggest removing 'throws Exception'.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/juniper/StatusKitRoute.java:
```diff
@@ -0,0 +1,59 @@
+public class StatusKitRoute implements Route {
+    NonPepperStatusKitService nonPepperStatusKitService;
+
+    public StatusKitRoute(NonPepperStatusKitService nonPepperStatusKitService) {
```
Unclear why this is passed in rather than created here.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/juniper/StatusKitRoute.java:
```diff
@@ -0,0 +1,59 @@
+
+@Slf4j
+public class StatusKitRoute implements Route {
+    NonPepperStatusKitService nonPepperStatusKitService;
```
This should be private and final.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/EasyPostUtil.java:
```diff
@@ -295,4 +297,51 @@ public Address getAddress(String addressId) throws EasyPostException {
+     * @param juniperKitRequest the JuniperKitRequest with address to check
+     */
+
+    public boolean checkAddress(JuniperKitRequest juniperKitRequest, String phone) {
```
Typically this would take KitRequest instead of JuniperKitRequest, but I was surprised to see that KitRequest does not include obvious fields like last name. It feels like the KitRequest inheritance could be refactored to something that makes more sense, but I realize that's out of scope here.

pegahtah replied:
This is because a kitRequest from DSS only has a guid, we get the address later from ES based on that guid.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/EasyPostUtil.java:
```diff
@@ -295,4 +297,51 @@ public Address getAddress(String addressId) throws EasyPostException {
+     * An address is valid only if participant has shortId, first - and lastName, for Juniper shortId is the juniperParticipantId
+     *
+     * @param juniperKitRequest the JuniperKitRequest with address to check
+     */
```
Since you done most of the work here, an @returns should be included.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/EasyPostUtil.java:
```diff
@@ -295,4 +297,51 @@ public Address getAddress(String addressId) throws EasyPostException {
+            name += juniperKitRequest.getFirstName() + " ";
+        }
+        name += juniperKitRequest.getLastName();
+        if (juniperKitRequest.isSkipAddressValidation()) {
```
Is this test correct? It looks like the code is validating the address, but the test seems to be the opposite.

pegahtah replied:
I changed the name of createBroadAddress to `createAddressWithoutValidation` to match better to what it does , we still need easypost to create the address in their system and return an id, but we don't need easypost to validate it. However, based on my experience this step in easypost also does some checks and validations, but the bar's much lower here. There are still cases where we get an exception from easypost and we are not able to create the address. For example a random not real name for the city or country or a weird zipcode would still take us to L330. 

However the real validation is in line 337if the skipAddressValidation is not set

denniscunningham replied:
Thx for the explanation.
I think I understand now. The lack of symmetry between the two ways we do this is difficult to discern. IOW, in one case we call a utility method (createAddres...), and in the other we instantiate a class. That's strange, since a reader might expect one code path with a flag passed saying whether validation is needed.
Perhaps this is an opportunity to add a comment to each section saying what they do?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/EasyPostUtil.java:
```diff
@@ -295,4 +297,51 @@ public Address getAddress(String addressId) throws EasyPostException {
+                new DeliveryAddress(juniperKitRequest.getStreet1(), juniperKitRequest.getStreet2(), juniperKitRequest.getCity(),
+                        juniperKitRequest.getState(),
+                        juniperKitRequest.getPostalCode(), juniperKitRequest.getCountry(), name, phone);
+        deliveryAddress.validate();
```
I'm confused. How does `isSkipAddressValidation` fit in here? Also, the implementation of DeliveryAddress can lead to errors, since the caller first has to call validate before isValid. That logic should all be in isValid. Again, out of scope here, but something that should be cleaned up at some point.

pegahtah replied:
I agree with this. about your first question, if `isSkipAddressValidation` is set to true, we don't call `validate()`, just try to create an address in easypost without asking if it's valid.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/EasyPostUtil.java:
```diff
@@ -295,4 +297,51 @@ public Address getAddress(String addressId) throws EasyPostException {
+            juniperKitRequest.setEasypostAddressId(deliveryAddress.getId());
+            return true;
+        }
+        logger.info("Address is not valid " + juniperKitRequest.getShortId());
```
Perhaps DeliveryAddress should have a toString method that could be used here so the log would have more information for debugging?

pegahtah replied:
I think that's then considered PHI, we can check the address by the participant id

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/JuniperKitCreationTest.java:
```diff
@@ -1,43 +1,66 @@
-    public void beforeClass() {
-        TestHelper.setupDB();
-        cfg = TestHelper.cfg;
+    public void setupJuniperBefore() {
```
Just to be clear, this will run before each test, not the entire test class. Is that what you want?

pegahtah replied:
Sorry I think I missed these comments last night, I will change to BeforeClass then

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/JuniperKitCreationTest.java:
```diff
@@ -1,43 +1,66 @@
-        cfg = TestHelper.cfg;
+    public void setupJuniperBefore() {
+        JuniperSetupUtil.setupUpAJuniperInstance(instanceName, instanceGuid, "Juniper-Mock", bspPrefix);
+        JuniperSetupUtil.loadDSMConfig();
```
Unclear why this is needed.

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/JuniperSetupUtil.java:
```diff
@@ -0,0 +1,325 @@
+            } catch (SQLException ex) {
+                throw new RuntimeException("Error inserting juniper group in ddp_group ", ex);
+            }
+            try (PreparedStatement stmt = conn.prepareStatement(INSERT_JUNIPER_INSTANCE, Statement.RETURN_GENERATED_KEYS)) {
```
It probably makes sense to put each of these clauses in a separate method. That way it is easy to see that the delete stuff below matches the things created here.

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/NonPepperKitServiceTest.java:
```diff
@@ -0,0 +1,77 @@
+        Assert.assertNotNull(timeStamp);
+
+        Instant instant = Instant.ofEpochMilli(now).atZone(ZoneId.of("UTC")).toInstant();
+        assertEquals(instant.toString(), timeStamp);
```
Can this lead to flaky behavior? If this takes more than a millisecond to execute will the assertion fail?

pegahtah replied:
It is still using the `now` calculated above for all of the tests, not getting newer time

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/NonPepperKitServiceTest.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @Test
+    public void convertTimeStringIntoTimeStampTest() {
```
I'm not sure this test is necessary, or could be greatly reduced. The convertTimeStringIntoTimeStamp method checks input then calls a library function, so the only thing we ned to test is whether bad input is handled correctly.

pegahtah replied:
I am mostly trying to make sure the parsed Date from `nonPepperStatusKitService.convertTimeStringIntoTimeStamp` has the required format

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/NonPepperKitServiceTest.java:
```diff
@@ -0,0 +1,77 @@
+        map.put(4, userDto4);
+        map.put(5, userDto5);
+
+        String user1TestEmail = nonPepperStatusKitService.getUserEmailForFields("1", map);
```
See my comment on getUserEmailForFields.

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/JuniperKitStatusTest.java:
```diff
@@ -0,0 +1,188 @@
+ * -ea -Dconfig.file=[path to /pepper-apis/output-build-config/testing-inmemorydb.conf]
+ *  */
+
+@RunWith(MockitoJUnitRunner.class)
```
I am not familiar with this annotation. Why was it necessary?

pegahtah replied:
I found [this post](https://stackoverflow.com/questions/10806345/runwithmockitojunitrunner-class-vs-mockitoannotations-initmocksthis) about it before using it. I am still a beginner in using Mockito but this is supposed to automatically validate framework usage

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/JuniperKitStatusTest.java:
```diff
@@ -0,0 +1,188 @@
+        Assert.assertNotNull(
+                kitResponse.getKits().stream().filter(kitStatus -> kitStatus.getJuniperKitId().equals("JuniperTestKitId_" + rand))
+                        .findAny());
+        for (NonPepperKitStatusDto nonPepperKitStatusDto : kitResponse.getKits()) {
```
This code may never get executed when an assertion fails. There are a couple of ways to deal with that: a try around whatever is creating the kits with a finally clause, or keep track of the kits at the class level and run this code in the After method.

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/JuniperKitStatusTest.java:
```diff
@@ -0,0 +1,188 @@
+        Assert.assertNotEquals(0, kitResponse.getKits().size());
+        Assert.assertNotNull(
+                kitResponse.getKits().stream().filter(kitStatus -> kitStatus.getJuniperKitId().equals("JuniperTestKitId_" + rand))
+                        .findAny());
```
Ideally you would check that the details of the kits the test created.

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/JuniperKitStatusTest.java:
```diff
@@ -0,0 +1,188 @@
+                + "\"skipAddressValidation\":false,"
+                + "\"juniperStudyID\":\"Juniper-mock-guid\"}";
+
+        JuniperKitRequest mockJuniperKit = new Gson().fromJson(json, JuniperKitRequest.class);
```
I'm not sure this would be considered a mock. It's more like a fixture.

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/JuniperKitStatusTest.java:
```diff
@@ -0,0 +1,188 @@
+
+    @Test
+    public void statusByJuniperKitId() {
+        String participantId = "TEST_PARTICIPANT";
```
Since this part is identical to the previous test it can be moved to the class level.

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/JuniperKitStatusTest.java:
```diff
@@ -0,0 +1,188 @@
+        Assert.assertEquals(newKits.size(), oldkits.size() + 1);
+
+        StatusKitResponse kitResponse =
+                (StatusKitResponse) nonPepperStatusKitService.getKitsBasedOnJuniperKitId(mockJuniperKit.getJuniperKitId());
```
Actually it looks like most of this code is the same as above. Perhaps create methods for the setup and assertions, or pass in the parts that are different?

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/JuniperKitStatusTest.java:
```diff
@@ -0,0 +1,188 @@
+
+    @Test
+    public void statusByKitIdTest() {
+        String participantId = "TEST_PARTICIPANT";
```
See my comment above.

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/JuniperKitCreationTest.java:
```diff
@@ -1,43 +1,66 @@
-public class JuniperKitCreationTest {
     NonPepperKitCreationService nonPepperKitCreationService = new NonPepperKitCreationService();
-    private Config cfg;
+    final String instanceGuid = "Juniper-mock-guid";
```
Nit: Typically we would use the term 'test' instead of 'mock', since the latter has a very specific meaning.

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/JuniperKitCreationTest.java:
```diff
@@ -1,43 +1,66 @@
 
     @Test
-    @Ignore
     public void createNewMockJuniperKitTest() {
```
Typically test method names include the primary thing they are testing, like myMethodTest. That convention makes it easy for readers to understand the intent of the test without Java doc.
In this case is that createNonPepperKit? The term Mock does not make sense here unless we are testing a mock, per se.

denniscunningham commented on pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/juniperkits/JuniperKitCreationTest.java:
```diff
@@ -98,29 +115,25 @@ public void noJuniperKitIdTest() {
+            List<KitRequestShipping> newKits = KitRequestShipping.getKitRequestsByRealm(instanceName, "overview", kitType);
+            Assert.assertEquals(newKits.size(), oldkits.size());
+        } finally {
+            JuniperSetupUtil.deleteJuniperKit(mockJuniperKit.getJuniperKitId());
```
Nice! See my comment in JuniperKitStatusTest.

denniscunningham reviewed (changes requested):
Lots of good work here! See my comments for details.
I may not have covered everything but I think there is enough in my comments for you to provide updates to the PR. I can make another pass after you do that.

There are areas where I did not comment things that should be fixed but should be covered by the linters. Have you installed SonarLint? Are you seeing Sonar and Checkstyle warnings in IntelliJ?

pegahtah committed:
PEPPER-814 PR comments addressed

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/KitResponse.java:
```diff
@@ -1,10 +1,5 @@
-    public String errorMessage;
-    public String juniperKitId;
-    public Object value;
+
```
Does this class simply exist so callers can distinguish an error response from a non-error response?
I have not been a fan of this pattern, but now that you refactored it feels like a simple boolean `isError` flag on KitResponse<Error> would be more straightforward?

denniscunningham reviewed (changes requested):
It looks like you are now responding to several comments you missed yesterday?
If so, I left a few new comments but will wait for another commit before continuing.

pegahtah committed:
PEPPER-814 PR comments addressed in tests, also PEPPER-1023

pegahtah committed:
PEPPER-814 PR comments addressed in tests, addressed as many sonar errors as I could

pegahtah committed:
PEPPER-814 PR comments addressed in tests,refactored JuniperSetupUtil

pegahtah committed:
PEPPER-814 PR comments addressed, merged KitResponseError.java and StatusKitResponse.java in KitResponse and removed those two classes, removed unused and committed out code

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into PEPPER-841_statusKit

pegahtah committed:
PEPPER-814 fixed missed value

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/kit/KitStatusDao.java:
```diff
@@ -0,0 +1,173 @@
+                while (rs.next()) {
+                    list.add((NonPepperKitStatusDto) builder.build(rs, users));
+                }
+                dbVals.resultValue = rs;
```
You can remove this, as you have in the other similar methods in this class.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dto/kit/nonPepperKit/NonPepperKitStatusDto.java:
```diff
@@ -0,0 +1,188 @@
+
+    public static class Builder {
+
+        Boolean error;
```
I assume you wanted to make this private?

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/KitResponse.java:
```diff
@@ -1,10 +1,71 @@
+    private List<NonPepperKitStatusDto> kits;
+    private boolean isError;
+
+    public KitResponse makeKitResponseError(ErrorMessage errorMessage, String juniperKitId, Object value) {
```
I like your approach here, but this is typically implemented differently. As this stands now, the 'this' in each of these maker methods is confusing.
The technique I often see (and use) is to make each of these static, then in each of those methods create an instance at the beginning (could be a default constructor or a constructor with the common members of each maker), then fill in the instance fields, and return the new instance.

Does that make sense?

pegahtah replied:
I think I get what you mean, I will make the changes.

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/juniper/StatusKitRoute.java:
```diff
@@ -0,0 +1,71 @@
+        KitResponse kitResponse;
+        if (request.requestMethod().equals(RoutePath.RequestMethod.GET.toString())) {
+            if (request.url().contains(RoutePath.KIT_STATUS_STUDY)) {
+                String study = request.params(RequestParameter.STUDY);
```
Each of the request params needs to be checked, at least to ensure they exist.
You can probably create a helper method that takes a parameter name, checks for a value, and throws a DSMBadRequestException. (That exception will be caught by our exception handling and return a 400 and a message to the requester.)

denniscunningham replied:
Sorry, my comment was unclear. I'm concerned what you have here now does not provide what you want, but I could be misunderstanding.
If I'm reading this correctly, you will get bad request exceptions for any case but the first?
My original concern was not the conditional test, but the first line in each clause, where you get the parameter. If that parameter is not there -- which is a bad request -- then at some point you will get a null pointer exception or similar (and thus a 500), rather than reporting a bad request.
I was thinking of a helper like `String getParam(String paramName, Request request)` that would throw a bad request exception if the param was not present or empty, otherwise return the pararm. But of course you could just put that logic in-line, if you prefer.
Does that make sense?

pegahtah replied:
Oh you are correct! I will fix this

denniscunningham reviewed (commented):
Very nice!!
A few more questions/comments.

pegahtah committed:
PEPPER-814 added comments to EasyPostUtil, added "juniper_study" to DBConstants as a static String, changed makeKitResponseError and makeKitStatusResponse to static methods, added a helper method to StatusKitRoute

denniscunningham commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/nonpepperkit/NonPepperStatusKitService.java:
```diff
@@ -0,0 +1,103 @@
+            log.info(studyGuid + " is not a Juniper study!");
+            return KitResponse.makeKitResponseError(KitResponse.ErrorMessage.UNKNOWN_STUDY);
+        }
+        HashMap<Integer, UserDto> users = (HashMap<Integer, UserDto>) UserDao.selectAllUsers();
```
I thought I commented on this, but I don't see anything....
Ideally we would select users as pat of the `getKitsByInstance` query, rather than joining them in code. As things stand now I do not believe this is a problem, since the total number of users is low. But if the number of users was high, then materializing many unnecessary users would be inefficient, and might even blow through instance memory.
No need for a change now, but this is sort of an antipattern.

pegahtah replied:
I remember your comment, but I can't also find that discussion! I thought since we only have around 200 users in the DB, it makes more sense to do one select and add them in the java code, rather than joining the tables or performing 4 different selects. I am always concerned about memory for DSM too. I _believe_ it won't go out of memory for holding 200 objects in memory but do you think this should change?

pegahtah replied:
Juniper plans to call this endpoint by study name in a nightly job, and calls during the day will mostly be by kit id or participant id. I know that it's not wise to design the endpoint only around that, but I thought it is good to know for the sake of this discussion

denniscunningham replied:
Yup, I agree that for our current number of users this is ok (and this should probably be ok for even ~1000 users). Just pointing out the anti-pattern. Ideally we would have implemented the join in the DB, not in code, but this is fine for now.

pegahtah committed:
PEPPER-841 PR comments applied, fixed the problem in StatusKitRoute and renamed method to deleteJuniperInstanceAndSettings in JuniperSetupUtil

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into PEPPER-841_statusKit

pegahtah committed:
PEPPER-841 PR comments applied, renamed setupJuniperInstance to setupJuniperInstanceAndSettings in JuniperSetupUtil, fixed the return of error true in times of success in NonPepperKitCreationService and fixed url for KIT_STATUS_PARTICIPANT_ID

denniscunningham reviewed (approved):
Nice PR!!

pegahtah committed:
PEPPER-841 fix for tests failing

pegahtah committed:
PEPPER-841 fix for tests failing

pegahtah committed:
PEPPER-841 added logs for tests and fixed a bug in logging in NonPepperKitCreationService

pegahtah committed:
PEPPER-841 testing a theory

pegahtah committed:
PEPPER-841 testing a theory

pegahtah committed:
Testing a theory

pegahtah committed:
PEPPER-841 testing a theory of why tests do not pass
